zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
-            // If the state was downgraded, some heights might already be upgraded.
-            // (Since the upgraded format is added to the tip, the database can switch between
-            // lower and higher versions at any block.)
-            //
-            // Keep upgrading until the initial database has been upgraded,
-            // or this task is cancelled by a shutdown.
-            while upgrade_height <= initial_tip_height
-                && matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty))
-            {
-                // TODO: Do one format upgrade step here
-                //
-                // This fake step just shows how to access the database.
-                let _replace_me_ = upgrade_db.tip();
-
-                upgrade_height = (upgrade_height + 1).expect("task exits before maximum height");
+        // Start of a database upgrade task.
+
+        let version_for_pruning_trees =
+            Version::parse("25.1.0").expect("Hardcoded version string should be valid.");
+
+        // Check if we need to prune the note commitment trees in the database.
+        if older_disk_version < version_for_pruning_trees {
+            // Get network upgrade heights
+            let (&sapling_height, _) = NetworkUpgrade::activation_list(network)
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Sapling)
+                .expect("there should be sapling upgrade");
+            let (&orchard_height, _) = NetworkUpgrade::activation_list(network)
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Nu5)
+                .expect("there should be Nu5 upgrade");
+
+            // Delete duplicates before sapling and orchard heights
+            let (mut prev_sapling_tree, mut prev_orchard_tree) = {
+                let height = Height(1);
+                let mut batch = DiskWriteBatch::new();
+
+                batch.delete_range_sapling_tree(&upgrade_db, &height, &sapling_height);
+                batch.delete_range_orchard_tree(&upgrade_db, &height, &orchard_height);
+                upgrade_db
+                    .write_batch(batch)
+                    .expect("Deleting note commitment trees should always succeed.");
+
+                (
+                    upgrade_db.sapling_tree_by_height(&Height(0)),
+                    upgrade_db.orchard_tree_by_height(&Height(0)),
+                )
+            };
+
+            // Create an unbounded channel for reading note commitment trees
+            let (sapling_tree_tx, sapling_tree_rx) = mpsc::channel();
+
+            // Set up task for reading sapling note commitment trees
+            let db = upgrade_db.clone();
+            let should_cancel_flag = should_cancel_format_change.clone();
+            let sapling_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.sapling_tree_by_height_range(sapling_height..initial_tip_height)
```

Comment:
We might also want to check for a similar missed duplicate bug in the first new block that is committed to the finalized state after Zebra is started.

Reply:
Addressed in [cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf).

> We might also want to check for a similar missed duplicate bug in the first new block that is committed to the finalized state after Zebra is started.

I don't think this can happen because duplicate trees aren't committed to the database anymore: https://github.com/zcashfoundation/zebra/blob/ead854b587d9a64ef1ecd5c4971b3a4736c04a91/zebra-state/src/service/finalized_state/zebra_db/shielded.rs#L339-L356.